service/history/shard/context_impl.go

```diff
@@ -1026,66 +1116,205 @@ func (s *ContextImpl) rUnlock() {
 	s.rwLock.RUnlock()
 }
 
-func acquireShard(
-	shardItem *historyShardsItem,
-	closeCallback func(*historyShardsItem),
-) (*ContextImpl, error) {
-
-	var shardInfo *persistence.ShardInfoWithFailover
-
-	retryPolicy := backoff.NewExponentialRetryPolicy(50 * time.Millisecond)
-	retryPolicy.SetMaximumInterval(time.Second)
-	retryPolicy.SetExpirationInterval(5 * time.Second)
-
-	retryPredicate := func(err error) bool {
-		if common.IsPersistenceTransientError(err) {
-			return true
+func (s *ContextImpl) String() string {
+	// use memory address as shard context identity
+	return fmt.Sprintf("%p", s)
+}
+
+func (s *ContextImpl) transitionLocked(request contextRequest) {
```

Comment:
do we have to manage these states? any possibility to simplify?

Reply:
I spent a while on this and explored lots of alternatives. If you have a concrete suggestion please share.